cron/yearly/email-report.js

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env node
+
+process.env.PORT = 3066;
+
+import models, {sequelize} from '../../server/models';
+import _ from 'lodash';
+import Promise from 'bluebird';
+import { formatCurrency, formatArrayToString, getTier, formatCurrencyObject } from '../../server/lib/utils';
+import emailLib from '../../server/lib/email';
+
+const d = new Date;
+const startDate = new Date(`${d.getFullYear()}`);
+const endDate = new Date(`${d.getFullYear()+1}`);
+
+console.log("startDate", startDate, "endDate", endDate);
+
+const GetUserTransactionsQuery = `
+with "UserTransactions" as (
+  SELECT
+    "GroupId",
+    SUM("amountInTxnCurrency") as "amountInTxnCurrency",
+    MAX("txnCurrency") as "txnCurrency",
+    SUM("platformFeeInTxnCurrency") as "platformFeeInTxnCurrency",
+    MAX("hostFeeInTxnCurrency") as "hostFeeInTxnCurrency",
+    MAX("paymentProcessorFeeInTxnCurrency") as "paymentProcessorFeeInTxnCurrency"
+  FROM "Transactions"
+  WHERE "UserId"=:userid AND amount > 0 AND "deletedAt" IS NULL AND "PaymentMethodId" IS NOT NULL
+  GROUP BY "GroupId"
+)
+SELECT
+  ut.*,
+  host.username as "hostSlug",
+  CONCAT(host."firstName", ' ', host."lastName") as "hostName",
+  host.avatar as "hostLogo", host."twitterHandle" as "hostTwitterHandle", host.description as "hostDescription", host.mission as "hostMission",
+  g.slug, g.name, g.mission, g.logo, g."backgroundImage", g."twitterHandle", g.tiers, g.settings, g.data 
+FROM "UserTransactions" ut 
+LEFT JOIN "Groups" g ON ut."GroupId" = g.id
+LEFT JOIN "UserGroups" ug ON ut."GroupId" = ug."GroupId" AND ug.role='HOST'
+LEFT JOIN "Users" host ON ug."UserId" = host.id`;
+
+const buildTweet = (collectives, totalDonations) => {
+  let tweet;
+  const totalCollectives = Object.keys(collectives).length;
+  const collectivesNames = [], collectivesDonationsNames = [];
+  for (const slug in collectives) {
+    const collective = collectives[slug];
+    const collectiveName = (collective.twitterHandle) ? `@${collective.twitterHandle}` : collective.name;
+    collectivesNames.push(collectiveName);
+    collectivesDonationsNames.push(`${formatCurrency(collective.totalDonations, collective.currency)} to ${collectiveName}`);
+  }
+
+  if (totalCollectives > 2) {
+    tweet = `🎁 In 2016, I have contributed a total of ${formatCurrencyObject(totalDonations)} to ${totalCollectives} collectives`;
+    const listCollectives = formatArrayToString(collectivesNames);
+    if ( `${tweet}: ${listCollectives}`.length < 120 )
+      tweet = `${tweet}: ${listCollectives}`;
+  } else if (totalCollectives === 1) {
+    tweet = `🎁 In 2016, I have contributed ${collectivesDonationsNames[0].replace(' to ',' to the ')} collective`;
+  } else {
+    tweet = `🎁 In 2016, I have contributed ${formatArrayToString(collectivesDonationsNames)}'s collective`;
+  }
+
+  return tweet;
+}
+
+const processUser = (user) => {
```

Comment:
It's hard to follow this function. Can you leave comments with the logic in the future?